File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void reverseVector(vector<int> &v)
    {
        for(int l = 0, r = v.size() - 1; l < r; l++, r--)
        {
            swap(v[l], v[r]);
        }    
    }    
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (!root)
            return {};
        
        vector<vector<int>> result;
        std::queue<TreeNode *> q;
        
        q.push(root);
        bool l2r = false;
        
        while (q.empty() == false) {
            int lSize = q.size();
            vector<int> thisLevel(0, lSize);

            for( int i = 0; i < lSize; i++)
            {
                TreeNode *tmp = q.front();
                q.pop();
                thisLevel.push_back(tmp->val);
                
                if (tmp->left)
                    q.push(tmp->left);
                if (tmp->right)
                    q.push(tmp->right);    
            }
            
            if(l2r)
                reverseVector(thisLevel);
            l2r = !l2r;
            result.push_back(thisLevel);
        }
        
        return result;
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        
        result = []
        l2r = True
        
        q = collections.deque([root])
        while len(q) != 0:
            temp = []
            numNodes = len(q)
            for _ in range(numNodes):
                node = q.popleft()
                temp.append(node.val)
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
            if l2r is False:
                temp.reverse()
            l2r = not l2r
            result.append(temp)
            
        return result
            